packages/parse-cli-args/src/index.ts

```diff
@@ -84,6 +90,26 @@ export default async function parseCliArgs (
     return 'add'
   }
 
+  function getEscapeArgsWithSpecialCaseForRun () {
+    if (cmd !== 'run') {
+      return opts.escapeArgs
+    }
+
+    // We'd like everything after the run script's name to be passed to the
+    // script's argv itself. For example, "pnpm run echo --test" should pass
+    // "--test" to the "echo" script. This requires determining the script's
+    // name and declaring it as the "escape arg".
+    //
+    // The name of the run script is normally the second argument (ex: pnpm
+    // run foo), but can be pushed back by recursive commands (ex: pnpm
+    // recursive run foo) or becomes the first argument when the fallback
+    // command (ex: pnpm foo) is set to 'run'.
+    const indexOfRunScriptName = 1 +
+      (recursiveCommandUsed ? 1 : 0) +
+      (fallbackCommandUsed && opts.fallbackCommand === 'run' ? -1 : 0)
```

Comment:
The logic here might be a bit of a hacky way to do things. We could also:

- run `nopt` again to find the script name to use as the escape arg. That would result in 3 `nopt` calls here.
- add additional behavior in pnpm's `nopt` fork to dynamically determine whether something is an escape arg. I'm imagining a closure that gets passed the partial CLI parsing up to that point.